File: python/leetcode_grind/day_501.py
# https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency/description/
from collections import Counter 

class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        cnt = Counter()
        ans, l = 0, 0
        for r in range(len(nums)):
            cnt[nums[r]] += 1
            while l < r and cnt[nums[r]] > k:
                cnt[nums[l]] -= 1
                l+=1
            ans = max(ans, r - l + 1)
        return ans

File: java/leetcode_grind/Day501.java
package leetcode_grind;

import java.util.HashMap;
import java.util.Map;

public class Day501 {
    // https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency/description
    static class Solution1 {
        public int maxSubarrayLength(int[] nums, int k) {
            int n = nums.length;
            Map<Integer, Integer> frequency = new HashMap<>();
            int start = 0;
            int charsWithFreqOverK = 0;

            for (int end = 0; end < n; end++) {
                frequency.put(nums[end], frequency.getOrDefault(nums[end], 0) + 1);

                if (frequency.get(nums[end]) == k + 1) {
                    charsWithFreqOverK++;
                }

                if (charsWithFreqOverK > 0) {
                    frequency.put(nums[start], frequency.getOrDefault(nums[start], 0) - 1);
                    if (frequency.get(nums[start]) == k) {
                        charsWithFreqOverK--;
                    }
                    start++;
                }
            }
            return n - start;
        }
    }
}
